brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/sora/ValDistributionPluggableLogic.java

```diff
@@ -0,0 +1,194 @@
+  private static final String SORA_DOMAIN = "sora";
+  public static final String VAL_ASSET_ID = "val#" + SORA_DOMAIN;
+  private static final String VAL_AIRDROP_DESCRIPTION = "VAL airdrop";
+  private static final int VAL_PRECISION = 18;
+  private static final MathContext VAL_MATH_CONTEXT = new MathContext(
+      Integer.MAX_VALUE,
+      RoundingMode.DOWN
+  );
+  private final IrohaAPI irohaAPI;
+  private final String brvsAccountId;
+  private final KeyPair brvsKeypair;
+  private final String infoSetterAccount;
+  private final IrohaQueryHelper irohaQueryHelper;
+  private final RegisteredUsersStorage registeredUsersStorage;
+  private final BigDecimal totalProportionPool;
+
+  public ValDistributionPluggableLogic(
+      QueryAPI queryAPI,
+      String infoSetterAccount,
+      IrohaQueryHelper irohaQueryHelper,
+      RegisteredUsersStorage registeredUsersStorage,
+      String totalProportionPool) {
+    Objects.requireNonNull(queryAPI, "Query API must not be null");
+    if (StringUtils.isEmpty(infoSetterAccount)) {
+      throw new IllegalArgumentException("Info setter account must not be neither null nor empty");
+    }
+    Objects.requireNonNull(irohaQueryHelper, "IrohaQueryHelper must not be null");
+    Objects.requireNonNull(registeredUsersStorage, "RegisteredUsersStorage must not be null");
+    if (StringUtils.isEmpty(totalProportionPool)) {
+      throw new IllegalArgumentException("TotalProportionPool must not be neither null nor empty");
+    }
+
+    this.irohaAPI = queryAPI.getApi();
+    this.brvsAccountId = queryAPI.getAccountId();
+    this.brvsKeypair = queryAPI.getKeyPair();
+    this.infoSetterAccount = infoSetterAccount;
+    this.irohaQueryHelper = irohaQueryHelper;
+    this.registeredUsersStorage = registeredUsersStorage;
+    this.totalProportionPool = new BigDecimal(totalProportionPool);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public BigDecimal filterAndTransform(Block block) {
+    return block.getBlockV1().getPayload().getTransactionsList()
+        .stream()
+        .filter(transaction ->
+            transaction.getPayload().getReducedPayload().getCreatorAccountId()
+                .equals(infoSetterAccount)
+        )
+        .flatMap(transaction ->
+            transaction.getPayload().getReducedPayload().getCommandsList().stream()
+        )
+        .filter(Command::hasTransferAsset)
+        .map(Command::getTransferAsset)
+        .filter(transferAsset -> transferAsset.getSrcAccountId().equals(infoSetterAccount) &&
+            transferAsset.getDestAccountId().equals(brvsAccountId) &&
+            transferAsset.getAssetId().equals(VAL_ASSET_ID)
+        )
+        .map(TransferAsset::getAmount)
+        .map(BigDecimal::new)
+        .reduce(
+            BigDecimal.ZERO,
+            BigDecimal::add
+        );
+  }
+
+  @Override
+  protected void applyInternal(BigDecimal amountToDistribute) {
+    if (amountToDistribute.signum() != 1) {
```

Comment:
Is this condition possible? Iroha does not allow to transfer non positive amount of asset